sdc/rewrites/dataframe_constructor.py

```diff
@@ -0,0 +1,200 @@
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# *****************************************************************************
+
+
+from numba.rewrites import (register_rewrite, Rewrite)
+from numba.ir_utils import (guard, find_callname)
+from numba.ir import (Expr)
+from numba.extending import overload
+
+from sdc.rewrites.ir_utils import (find_operations, is_tuple, is_dict, is_list,
+                                    get_tuple_items, get_dict_items, remove_unused,
+                                    get_call_parameters,
+                                    declare_constant,
+                                    import_function, make_call,
+                                    insert_before)
+from sdc.hiframes.pd_dataframe_ext import (init_dataframe)
+
+from sdc.hiframes.api import fix_df_array
+
+from pandas import DataFrame
+
+
+@register_rewrite('before-inference')
+class RewriteDataFrame(Rewrite):
+    """
+    Searches for calls of pandas.DataFrame and replace it with calls of init_dataframe.
+    Supports two cases:
+        1. data and columns parameters are tuples
+        2. data parameter is dictionary
+    """
+
+    _pandas_dataframe = ('DataFrame', 'pandas')
+
+    _df_arg_list = ('data', 'index', 'columns', 'dtype', 'copy')
+
+    def __init__(self, pipeline):
+        super().__init__(pipeline)
+
+        self._reset()
+
+
+    def match(self, func_ir, block, typemap, calltypes):
+        self._reset()
+
+        self._block = block
+        self._func_ir = func_ir
+        self._calls_to_rewrite = set()
+
+        for stmt in find_operations(block=block, op_name='call'):
+            expr = stmt.value
+            fdef = guard(find_callname, func_ir, expr)
+            if fdef == self._pandas_dataframe:
+                args = get_call_parameters(call=expr, arg_names=self._df_arg_list)
+
+                if self._match_tuple_case(args, block, func_ir):
+                    self._calls_to_rewrite.add((stmt, self._extract_tuple_args, lambda *args: None))
+                elif self._match_dict_case(args, block, func_ir):
+                    self._calls_to_rewrite.add((stmt, self._extract_dict_args, self._remove_dict_var))
+                else:
+                    pass #TODO: throw error?
+
+        return len(self._calls_to_rewrite) > 0
+
+
+    def apply(self):
+        init_df_stmt = import_function(init_dataframe, self._block, self._func_ir)
+
+        for stmt, extract, cleanup in self._calls_to_rewrite:
+            args = get_call_parameters(call=stmt.value, arg_names=self._df_arg_list)
+
+            old_data = args['data']
+            old_columns = args['columns'] if 'columns' in args else None
```

Comment:
Looks like this line is equal to `old_columns = args.get('columns')`.

Reply:
Thanks, fixed